### PR TITLE
fix: follow symlinks

### LINKED ISSUE #1823: List API should follow symlinks

### PR DESCRIPTION
**Pull request by andrewrynhard**
This fixes the list API to check if the requested path is a symlink, and
to follow the symlink if so.

Fixes #1823.

### CODE REVIEW

**Comment by andrewrynhard** [comment]
```
$ osctl ls /proc/thread-self -l
NODE            MODE         SIZE(B)   LASTMOD           NAME
147.75.64.133   dr-xr-xr-x   0         Jan 28 08:57:25   .
147.75.64.133   -r--r--r--   0         Jan 28 08:57:25   arch_status
147.75.64.133   dr-xr-xr-x   0         Jan 28 08:57:25   attr
147.75.64.133   -r--------   0         Jan 28 08:57:25   auxv
147.75.64.133   -r--r--r--   0         Jan 28 08:57:25   cgroup
147.75.64.133   --w-------   0         Jan 28 08:57:25   clear_refs
147.75.64.133   -r--r--r--   0         Jan 28 08:57:25   cmdline
147.75.64.133   -rw-r--r--   0         Jan 28 08:57:25   comm
147.75.64.133   -r--r--r--   0         Jan 28 08:57:25   cpuset
147.75.64.133   Lrwxrwxrwx   0         Jan 28 08:57:25   cwd -> /
147.75.64.133   -r--------   0         Jan 28 08:57:25   environ
147.75.64.133   Lrwxrwxrwx   0         Jan 28 08:57:25   exe -> /sbin/init
147.75.64.133   dr-x------   0         Jan 28 08:57:25   fd
147.75.64.133   dr-x------   0         Jan 28 08:57:25   fdinfo
147.75.64.133   -rw-r--r--   0         Jan 28 08:57:25   gid_map
147.75.64.133   -r--------   0         Jan 28 08:57:25   io
147.75.64.133   -r--r--r--   0         Jan 28 08:57:25   limits
147.75.64.133   -rw-r--r--   0         Jan 28 08:57:25   loginuid
147.75.64.133   -r--r--r--   0         Jan 28 08:57:25   maps
147.75.64.133   -rw-------   0         Jan 28 08:57:25   mem
147.75.64.133   -r--r--r--   0         Jan 28 08:57:25   mountinfo
147.75.64.133   -r--r--r--   0         Jan 28 08:57:25   mounts
147.75.64.133   dr-xr-xr-x   0         Jan 28 08:57:25   net
147.75.64.133   dr-x--x--x   0         Jan 28 08:57:25   ns
147.75.64.133   -r--r--r--   0         Jan 28 08:57:25   numa_maps
147.75.64.133   -rw-r--r--   0         Jan 28 08:57:25   oom_adj
147.75.64.133   -r--r--r--   0         Jan 28 08:57:25   oom_score
147.75.64.133   -rw-r--r--   0         Jan 28 08:57:25   oom_score_adj
147.75.64.133   -r--------   0         Jan 28 08:57:25   pagemap
147.75.64.133   -r--------   0         Jan 28 08:57:25   personality
147.75.64.133   -rw-r--r--   0         Jan 28 08:57:25   projid_map
147.75.64.133   Lrwxrwxrwx   0         Jan 28 08:57:25   root -> /
147.75.64.133   -r--r--r--   0         Jan 28 08:57:25   schedstat
147.75.64.133   -r--r--r--   0         Jan 28 08:57:25   sessionid
147.75.64.133   -rw-r--r--   0         Jan 28 08:57:25   setgroups
147.75.64.133   -r--r--r--   0         Jan 28 08:57:25   smaps
147.75.64.133   -r--r--r--   0         Jan 28 08:57:25   smaps_rollup
147.75.64.133   -r--------   0         Jan 28 08:57:25   stack
147.75.64.133   -r--r--r--   0         Jan 28 08:57:25   stat
147.75.64.133   -r--r--r--   0         Jan 28 08:57:25   statm
147.75.64.133   -r--r--r--   0         Jan 28 08:57:25   status
147.75.64.133   -r--------   0         Jan 28 08:57:25   syscall
147.75.64.133   -rw-r--r--   0         Jan 28 08:57:25   uid_map
147.75.64.133   -r--r--r--   0         Jan 28 08:57:25   wchan
$ osctl ls /proc/thread-self/exe -l
NODE            MODE         SIZE(B)    LASTMOD           NAME
147.75.64.133   -rwxr-xr-x   47538552   Jan 28 08:49:50   init
```

File: pkg/archiver/walker.go [hunk 1]
```diff
@@ -55,11 +55,18 @@ func Walker(ctx context.Context, rootPath string, options ...WalkerOption) (<-ch
...
 	}
 
-	_, err := os.Stat(rootPath)
+	info, err := os.Lstat(rootPath)
 	if err != nil {
 		return nil, err
 	}
 
+	if info.Mode()&os.ModeSymlink == os.ModeSymlink {
+		rootPath, err = filepath.EvalSymlinks(rootPath)
```
Comment (smira) [inline pkg/archiver/walker.go]: as a suggestion, should we just use `os.Readlink` (https://godoc.org/os#Readlink) here, as it seems to be closer to what we're looking for? `EvalSymlinks` should do more than just final destination?

**Comment by smira** [comment]
I feel it's a perfect place to have a test ;)

File: pkg/archiver/walker.go (hunk 1 above)
Comment (smira) [reply pkg/archiver/walker.go]: sorry, was my misunderstanding

**Pull request merged**